### PR TITLE
GORA-266 Lucene datastore for Gora

### PR DESCRIPTION
**Pull request by lewismc**
This PR updates the patch at https://issues.apache.org/jira/browse/GORA-266 and also upgrades the Lucene API's to match with 6.2.1 which is currently where the Solr/Lucene dependencies are pegged in parent pom.xml.

Some issues with this PR

1. tests do not pass
1. UNION types are not supported
1. getPartitions is not supported
1. There are a number of inline comments which we should address

Code review welcome, thanks

### CODE REVIEW

**Comment by renato2099** [comment]
Hey @lewismc 
This looks awesome! Thanks a lot for doing this! I went over the code quickly and I left some questions and nitpicks, feel free to update whenever you have time and ping me back if you want another round of reviews.

**Comment by lewismc** [comment]
Thanks @renato2099 where are the comments? I don't see them.

**Comment by renato2099** [comment]
if you look at the file committed then you should see them. Although I also see them directly here in the PR page.

**Comment by renato2099** [comment]
sorry @lewismc I didn't submitted, I just did now.

File: gora-core/src/main/java/org/apache/gora/util/IOUtils.java [hunk 1]
```diff
@@ -434,7 +432,7 @@ public static void writeBoolArray(DataOutput out, boolean[] boolArray)
     WritableUtils.writeVInt(out, boolArray.length);
 
     byte b = 0;
-    int i = 0;
+    int i;
```
[the remark this answers is not in the document]
Comment (lewismc) [reply gora-core/src/main/java/org/apache/gora/util/IOUtils.java]: You mean an enhanced for loop? From what I can see the variable ```i``` is used outside of the ```for(i=0; i<boolArray.length; i++) {``` statement in line 447 at ```if(i % 8 != 0) {```

[the remark this answers is not in the document]
Comment (renato2099) [reply gora-core/src/main/java/org/apache/gora/util/IOUtils.java]: yeah that is what I meant.
for(int i=0; i<boolArray.length; i++) {

[the remark this answers is not in the document]
Comment (lewismc) [reply gora-core/src/main/java/org/apache/gora/util/IOUtils.java]: @renato2099 as I said the increment is used in line 447

[the remark this answers is not in the document]
Comment (renato2099) [reply gora-core/src/main/java/org/apache/gora/util/IOUtils.java]: right, I thought that if was inside the for loop.
Sorry for the noise on this one.

File: gora-core/src/main/java/org/apache/gora/util/IOUtils.java [hunk 2]
```diff
@@ -660,7 +657,8 @@ public static void writeStringArray(DataOutput out, String[] arr)
         buffer.limit(count);
         buffers.add(buffer);
       }
-      if(count < BUFFER_SIZE) break;
+      if(count < BUFFER_SIZE)
```
Comment (djkevincr) [inline gora-core/src/main/java/org/apache/gora/util/IOUtils.java]: @lewismc if ( - Seems this code is not formatted properly. May be these were coming from the existing code. Can you please check?

**Comment by djkevincr** [comment]
@lewismc Are you happy with this PR or is there anything else that we should address or do you need further time on this? Its awesome if we can get to the release coming shortly.

**Comment by lewismc** [comment]
I don't have time to go back right now and code this implementation. I would suggest the following
* upgrade Lucene and fix any deprecated API usage
* fix the conflicts
* implement tests if they have no already be done.

**Comment by djkevincr** [comment]
@lewismc Thanks for the input Lewis, I will have look and take over this.

**Comment by djkevincr** [comment]
Closing this PR after submitting more update version of the same work in https://github.com/apache/gora/pull/152

**Pull request closed without merging**